### PR TITLE
[DDK] RtlUnicodeStringPrintf(Ex) implementations

### PR DESCRIPTION
**Pull request by hernandp**
## Purpose

Implement RtlUnicodeStringPrintf(Ex) functions for DDK String Safe functions. 

JIRA issue: [CORE-9819](https://jira.reactos.org/browse/CORE-9819)

## Proposed changes

First implementation (working) of those functions. For a bunch of elementary testing, see work being done on  [CORE-14565](https://jira.reactos.org/browse/CORE-14565)

Existing xxxStringPrintfW/xxxStringPrintfWEx functions call now an intermediate private function, but should not affect existing callers.

For details see last commit notes.

## TODO

- More test cases involving (Ex) function flags and boundary conditions.

### CODE REVIEW

File: sdk/include/ddk/ntstrsafe.h [hunk 1]
```diff
@@ -2825,6 +2826,45 @@ RtlStringLengthWorkerW(
     return Status;
 }
 
+NTSTRSAFEAPI 
+RtlUnicodeStringPrintf(PUNICODE_STRING DestinationString,
+	NTSTRSAFE_PCWSTR pszFormat, ...)
```
Comment (HBelusca) [inline sdk/include/ddk/ntstrsafe.h]: Please use correct coding style everywhere in your code:
- 4-space indentation;
- optionally for "if" blocks with 1-line case, surround around braces;
- The prototypes of the functions should be like:
```c
NTSTRSAFEAPI
RtlFunctionName(
  <SAL annotation> PUNICODE_STRING Parameter,
  <SAL annotation> TYPE OtherParameter)
```

File: sdk/include/ddk/ntstrsafe.h [hunk 2]
```diff
@@ -2825,6 +2826,45 @@ RtlStringLengthWorkerW(
...
+{
+	if (DestinationString == NULL || pszFormat == NULL)
+		return STATUS_INVALID_PARAMETER;
+
+	if (DestinationString->Buffer == NULL || DestinationString->MaximumLength == 0)
+		return STATUS_INVALID_PARAMETER;
+	
+	size_t cchLength = 0;
```
Comment (HBelusca) [inline sdk/include/ddk/ntstrsafe.h]: This won't compile, you must move the variable declaration at the beginning of the function.

File: sdk/include/ddk/ntstrsafe.h [hunk 3]
```diff
@@ -2825,6 +2826,45 @@ RtlStringLengthWorkerW(
...
+	NTSTATUS nt = RtlStringLengthWorkerW(DestinationString->Length, NTSTRSAFE_UNICODE_STRING_MAX_CCH, &cchLength);
```
Comment (HBelusca) [inline sdk/include/ddk/ntstrsafe.h]: Same comment for this one. Also, status variables are called "Status", not "nt".

File: sdk/include/ddk/ntstrsafe.h [hunk 4]
```diff
@@ -2825,6 +2826,45 @@ RtlStringLengthWorkerW(
...
+	ASSERT(NT_SUCCESS(nt));
```
Comment (HBelusca) [inline sdk/include/ddk/ntstrsafe.h]: NT_VERIFY ?

File: sdk/include/ddk/ntstrsafe.h [hunk 5]
```diff
@@ -2825,6 +2826,45 @@ RtlStringLengthWorkerW(
...
+	
+	/* Don't allow inconsistency between actual length and UNICODE_STRING.Length field, 
+	 * or "full buffer" (MaximumLength == Length)
+	 */
+	
+	if ( ((DestinationString->Length == 0) && (cchLength != 0)) ||
+		 ((DestinationString->Length == DestinationString->MaximumLength)))
+		return STATUS_INVALID_PARAMETER;
+
+	return RtlStringVPrintfWorkerW(DestinationString->Buffer, DestinationString->MaximumLength, pszFormat, argList);
```
Comment (HBelusca) [inline sdk/include/ddk/ntstrsafe.h]: If code lines are too long, you can do:
```
return FunctionName(FirstParam,
                    SecondParam,
                    etc...);
```

File: sdk/include/ddk/ntstrsafe.h [hunk 6]
```diff
@@ -2825,6 +2826,45 @@ RtlStringLengthWorkerW(
...
+	PUNICODE_STRING RemainingString,
+	DWORD dwFlags,
+	NTSTRSAFE_PCWSTR pszFormat, ...)
+{
+	/* Verify flags */
+
+	if (dwFlags == STRSAFE_FILL_BEHIND)
+	{
+
+	}
```
Comment (HBelusca) [inline sdk/include/ddk/ntstrsafe.h]: I think that here you can just call RtlStringPrintfExW with the correct parameters (the UNICODE_STRING already provide the needed buffer pointers and their buffer lenghts).

File: sdk/include/ddk/ntstrsafe.h [hunk 7]
```diff
@@ -2825,6 +2826,45 @@ RtlStringLengthWorkerW(
...
+	if (DestinationString->Buffer == NULL || DestinationString->MaximumLength == 0)
+		return STATUS_INVALID_PARAMETER;
+	
+	size_t cchLength = 0;
+	NTSTATUS nt = RtlStringLengthWorkerW(DestinationString->Length, NTSTRSAFE_UNICODE_STRING_MAX_CCH, &cchLength);
+	ASSERT(NT_SUCCESS(nt));
+	
+	/* Don't allow inconsistency between actual length and UNICODE_STRING.Length field, 
+	 * or "full buffer" (MaximumLength == Length)
+	 */
```
Comment (HBelusca) [inline sdk/include/ddk/ntstrsafe.h]: Suggestion, maybe using the `RtlValidateUnicodeString()` function?

File: sdk/include/ddk/ntstrsafe.h [hunk 8]
```diff
@@ -58,6 +58,9 @@
 #define STRSAFE_GET_FILL_PATTERN(dwFlags) ((int)((dwFlags) & 0x000000FF))
 #endif
 
+#define STRSAFE_FILL_BEHIND                 STRSAFE_FILL_BEHIND_NULL
+#define STRSAFE_ZERO_LENGTH_ON_FAILURE      STRSAFE_NULL_ON_FAILURE
```
Comment (HBelusca) [inline sdk/include/ddk/ntstrsafe.h]: What about moving these two flags upper, close the other ones?

File: sdk/include/ddk/ntstrsafe.h [hunk 9]
```diff
@@ -2534,33 +2540,48 @@ NTSTRSAFEAPI RtlStringVPrintfWorkerA(STRSAFE_LPSTR pszDest,size_t cchDest,STRSAF
...
-        int iRet;
-        size_t cchMax;
-        cchMax = cchDest - 1;
-        iRet = _vsnwprintf(pszDest,cchMax,pszFormat,argList);
-        if ((iRet < 0) || (((size_t)iRet) > cchMax))
+        pszDest += cchMax;
+        *pszDest = L'\0';
+        Status = STATUS_BUFFER_OVERFLOW;
+    }
+    else
```
Comment (HBelusca) [inline sdk/include/ddk/ntstrsafe.h]: The block inside the `else` needs its enclosing brackets.

File: sdk/include/ddk/ntstrsafe.h [hunk 10]
```diff
@@ -2534,33 +2540,48 @@ NTSTRSAFEAPI RtlStringVPrintfWorkerA(STRSAFE_LPSTR pszDest,size_t cchDest,STRSAF
...
+    iRet = _vsnwprintf(pszDest, cchMax, pszFormat, argList);
+    if ((iRet < 0) || (((size_t) iRet) > cchMax))
     {
-        int iRet;
-        size_t cchMax;
-        cchMax = cchDest - 1;
-        iRet = _vsnwprintf(pszDest,cchMax,pszFormat,argList);
-        if ((iRet < 0) || (((size_t)iRet) > cchMax))
+        pszDest += cchMax;
+        *pszDest = L'\0';
```
Comment (HBelusca) [inline sdk/include/ddk/ntstrsafe.h]: UNICODE_NULL

File: sdk/include/ddk/ntstrsafe.h [hunk 11]
```diff
@@ -2534,33 +2540,48 @@ NTSTRSAFEAPI RtlStringVPrintfWorkerA(STRSAFE_LPSTR pszDest,size_t cchDest,STRSAF
...
+        Status = STATUS_BUFFER_OVERFLOW;
+    }
+    else
+        if (((size_t) iRet) == cchMax)
         {
             pszDest += cchMax;
             *pszDest = L'\0';
```
Comment (HBelusca) [inline sdk/include/ddk/ntstrsafe.h]: UNICODE_NULL

File: sdk/include/ddk/ntstrsafe.h [hunk 12]
```diff
@@ -2825,6 +2885,121 @@ RtlStringLengthWorkerW(
...
+        (SourceString->Buffer == NULL && (SourceString->Length != 0 || SourceString->MaximumLength != 0)))
+    {
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    return STATUS_SUCCESS;
+}
+
+NTSTRSAFEAPI
+RtlpUnicodeStringValidateEx(PCUNICODE_STRING SourceString, STRSAFE_DWORD dwFlags)
```
Comment (HBelusca) [inline sdk/include/ddk/ntstrsafe.h]: Isn't it ccalled "[RtlUnicodeStringValidateEx](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/content/ntstrsafe/nf-ntstrsafe-rtlunicodestringvalidateex)" without the 'p'?

File: sdk/include/ddk/ntstrsafe.h [hunk 13]
```diff
@@ -2825,6 +2885,121 @@ RtlStringLengthWorkerW(
...
+{
+    if (dwFlags & STRSAFE_IGNORE_NULLS)
+        return STATUS_SUCCESS;
+
+    return RtlUnicodeStringValidate(SourceString);
```
Comment (HBelusca) [inline sdk/include/ddk/ntstrsafe.h]: It might be better to actually implement the RtlUnicodeStringValidate() as a special case of RtlUnicodeStringValidateEx().
Also, could you check whether all these may not be particular cases of the "[RtlValidateUnicodeString()](https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/lib/rtl/unicode.c;hb=579a784e0431e5400dacaf39ee6d97b0a380ebe9#l2544)" function?

**Comment by HBelusca** [comment]
At some point you will have to amend the author of the first of your commits so that when the PR will be ready for merging all the information is correct.

File: sdk/include/ddk/ntstrsafe.h (hunk 8 above)
Comment (hernandp) [reply sdk/include/ddk/ntstrsafe.h]: Agreed.

File: sdk/include/ddk/ntstrsafe.h (hunk 9 above)
Comment (hernandp) [reply sdk/include/ddk/ntstrsafe.h]: This is part of original code. My intention was not to modify existing functions.  Shall I touch this?

File: sdk/include/ddk/ntstrsafe.h (hunk 10 above)
Comment (hernandp) [reply sdk/include/ddk/ntstrsafe.h]: Same: This is part of original code. My intention was not to modify existing functions.  Shall I touch this?

File: sdk/include/ddk/ntstrsafe.h (hunk 11 above)
Comment (hernandp) [reply sdk/include/ddk/ntstrsafe.h]: Same : This is part of original code. My intention was not to modify existing functions.  Shall I touch this?

File: sdk/include/ddk/ntstrsafe.h (hunk 13 above)
Comment (hernandp) [reply sdk/include/ddk/ntstrsafe.h]: I will review this also.

File: sdk/include/ddk/ntstrsafe.h (hunk 12 above)
Comment (hernandp) [reply sdk/include/ddk/ntstrsafe.h]: Thanks for the tip, I didn't realize the ValidateEx existed as a public API.

File: sdk/include/ddk/ntstrsafe.h (hunk 9 above)
Comment (HBelusca) [reply sdk/include/ddk/ntstrsafe.h]: For this one it's ok to touch it.

File: sdk/include/ddk/ntstrsafe.h (hunk 10 above)
Comment (HBelusca) [reply sdk/include/ddk/ntstrsafe.h]: Ok, keep it as it is.

File: sdk/include/ddk/ntstrsafe.h (hunk 11 above)
Comment (HBelusca) [reply sdk/include/ddk/ntstrsafe.h]: Ditto.

File: sdk/include/ddk/ntstrsafe.h [hunk 14]
```diff
@@ -2534,33 +2540,48 @@ NTSTRSAFEAPI RtlStringVPrintfWorkerA(STRSAFE_LPSTR pszDest,size_t cchDest,STRSAF
...
-            }
-    }
+
+    if (pcchDestNewLen)
+        *pcchDestNewLen = (iRet == -1) ? cchDest : iRet;
+
     return Status;
 }
 
+NTSTRSAFEAPI RtlStringVPrintfWorkerW(STRSAFE_LPWSTR pszDest,size_t cchDest,STRSAFE_LPCWSTR pszFormat,va_list argList)
```
Comment (Fraizeraust) [inline sdk/include/ddk/ntstrsafe.h]: You might want to indent these long parameters of `RtlStringVPrintfWorkerW` function.

File: sdk/include/ddk/ntstrsafe.h [hunk 15]
```diff
@@ -2534,33 +2540,48 @@ NTSTRSAFEAPI RtlStringVPrintfWorkerA(STRSAFE_LPSTR pszDest,size_t cchDest,STRSAF
...
+{
+    if (cchDest==0)
+        return STATUS_INVALID_PARAMETER;
+
+    return RtlpStringVPrintfWorkerW(pszDest, cchDest, pszFormat, NULL, argList);
+}
+
+NTSTRSAFEAPI RtlStringVPrintfWorkerLenW(STRSAFE_LPWSTR pszDest, size_t cchDest, STRSAFE_LPCWSTR pszFormat, size_t* pcchDestNewLen, va_list argList)
```
Comment (Fraizeraust) [inline sdk/include/ddk/ntstrsafe.h]: Ditto.

File: sdk/include/ddk/ntstrsafe.h [hunk 16]
```diff
@@ -2667,8 +2688,15 @@ NTSTRSAFEAPI RtlStringVPrintfExWorkerA(STRSAFE_LPSTR pszDest,size_t cchDest,size
...
-NTSTRSAFEAPI RtlStringVPrintfExWorkerW(STRSAFE_LPWSTR pszDest,size_t cchDest,size_t cbDest,STRSAFE_LPWSTR *ppszDestEnd,size_t *pcchRemaining,STRSAFE_DWORD dwFlags,STRSAFE_LPCWSTR pszFormat,va_list argList)
+NTSTRSAFEAPI RtlpStringVPrintfExWorkerW(STRSAFE_LPWSTR pszDest,
+    size_t cchDest,
+    size_t cbDest,
+    STRSAFE_LPWSTR *ppszDestEnd,
+    size_t *pcchRemaining,
+    STRSAFE_DWORD dwFlags,
+    STRSAFE_LPCWSTR pszFormat,
+    size_t* pcchDestNewLen, 
+    va_list argList)
```
Comment (Fraizeraust) [inline sdk/include/ddk/ntstrsafe.h]: More indentation needed for these parameters. 

https://imgur.com/a/9hV88UG

File: sdk/include/ddk/ntstrsafe.h [hunk 17]
```diff
@@ -2775,6 +2806,35 @@ NTSTRSAFEAPI RtlStringVPrintfExWorkerW(STRSAFE_LPWSTR pszDest,size_t cchDest,siz
...
 }
 
+NTSTRSAFEAPI RtlStringVPrintfExWorkerW(STRSAFE_LPWSTR pszDest, 
+    size_t cchDest, 
+    size_t cbDest, 
+    STRSAFE_LPWSTR *ppszDestEnd, 
+    size_t *pcchRemaining, 
+    STRSAFE_DWORD dwFlags, 
+    STRSAFE_LPCWSTR pszFormat, 
+    va_list argList)
```
Comment (Fraizeraust) [inline sdk/include/ddk/ntstrsafe.h]: Ditto.

File: sdk/include/ddk/ntstrsafe.h [hunk 18]
```diff
@@ -2775,6 +2806,35 @@ NTSTRSAFEAPI RtlStringVPrintfExWorkerW(STRSAFE_LPWSTR pszDest,size_t cchDest,siz
...
+
+NTSTRSAFEAPI RtlStringVPrintfExWorkerLenW(STRSAFE_LPWSTR pszDest, 
+    size_t cchDest, 
+    size_t cbDest, 
+    STRSAFE_LPWSTR *ppszDestEnd, 
+    size_t *pcchRemaining, 
+    STRSAFE_DWORD dwFlags, 
+    STRSAFE_LPCWSTR pszFormat, 
+    size_t* pcchDestNewLen, 
+    va_list argList)
```
Comment (Fraizeraust) [inline sdk/include/ddk/ntstrsafe.h]: Ditto.

**Comment by ColinFinck** [comment]
Work will be continued in https://github.com/reactos/reactos/pull/1859
I'll leave this PR open to resolve conversations one by one until PR#1859 is merged.

File: sdk/include/ddk/ntstrsafe.h (hunk 16 above)
Comment (ColinFinck) [reply sdk/include/ddk/ntstrsafe.h]: Not necessarily, that's just preference.
I also prefer to indent parameters always with 4 spaces instead of aligning them to the function name.

File: sdk/include/ddk/ntstrsafe.h (hunk 11 above)
Comment (ColinFinck) [reply sdk/include/ddk/ntstrsafe.h]: Using `ANSI_NULL` or `UNICODE_NULL` requires including an NT header that defines them, which is something that neither we nor the original MS ntstrsafe.h do right now.
So let's not fix what isn't broken.

**Comment by ColinFinck** [comment]
PR #1859 has been merged!
